line by line logging

avikchaudhuri opened the pull request:
Summary:
Today there is no good mechanism to detect progress of non-strict export line-by-line in user code. This caused some pain recently in trying to find the exact line of user code that was triggering a bug where the process appeared stuck because deep down something was calling some symbolic shapes code that was suffering some exponential blowup.

This PR adds a environment variable for extended debugging that will log the line of user code corresponding to every torch function call. It only works in non-strict export for now. Prefix setting this environment variable with `TORCH_LOGS`  enabled for `export` logs at `DEBUG` level (i.e., with a `+` prefix), i.e.,.:

```
TORCHEXPORT_EXTENDED_DEBUG_CURRENT_LOC=1 TORCH_LOGS="+export" ...
```

This will show logs with something like:
```
...
prim::device called at .../example.py:4284 in foo
TensorBase.item called at .../example.py:4277 in bar
...
```

We already have an existing place to intercept torch functions where we process data-dependent errors in non-strict, so parking the logging there. An alternative place we could be doing this is where we add `stack_trace` metadata when generating code, but unfortunately at least the example that motivated this gets stuck before generating code, so that would be too late.

Test Plan: ran it on some sample commands

Differential Revision: D61692156

angelayi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5547,31 +5558,28 @@ def _suggest_fixes_for_data_dependent_error_non_strict(e):
+        _suggest_torch_checks(e, src_map)
 
 
 class _DataDependentErrorHandlerNonStrict(torch.overrides.TorchFunctionMode):
```
nit: wondering if it would be a good idea to move this into export code instead of cluttering up symbolic_shapes.py? We could also rename it to just be more generic, like "NonStrictDebugHandler" since now it handles both data-dependent error handling, and the extra debug logging now.

avikchaudhuri committed:
line by line logging (#134298)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134298

Today there is no good mechanism to detect progress of non-strict export line-by-line in user code. This caused some pain recently in trying to find the exact line of user code that was triggering a bug where the process appeared stuck because deep down something was calling some symbolic shapes code that was suffering some exponential blowup.

This PR adds a environment variable for extended debugging that will log the line of user code corresponding to every torch function call. It only works in non-strict export for now. Prefix setting this environment variable with `TORCH_LOGS` level of `DEBUG` (i.e., with a `+` prefix) for `dynamic` or `export` logs, e.g.:

```
TORCHEXPORT_EXTENDED_DEBUG_CURRENT_LOC=1 TORCH_LOGS="+export" ...
```

This will show logs with something like:
```
...
prim::device called at .../example.py:4284 in foo
TensorBase.item called at .../example.py:4277 in bar
...
```

We already have an existing place to intercept torch functions where we process data-dependent errors in non-strict, so parking the logging there. An alternative place we could be doing this is where we add `stack_trace` metadata when generating code, but unfortunately at least the example that motivated this gets stuck before generating code, so that would be too late.

Test Plan: ran it on some sample commands

Reviewed By: angelayi

Differential Revision: D61692156
